### PR TITLE
MM 47849 - tourtip after template creation

### PR DESCRIPTION
**Pull request by pvev**
#### Summary
This PR adds the necessary code to show the tourtip after a template has been used highlighting the product items that were created as part of the channel (boards and playbooks)

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-47849

#### Related Pull Requests
n/a

#### Screenshots

https://user-images.githubusercontent.com/10082627/219499009-ca7649be-9b1c-47e2-89fd-a97bd1f1f1a0.mov

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: components/tours/hooks.ts [hunk 1]
```diff
@@ -5,44 +5,72 @@ import {useCallback} from 'react';
...
+    const pluginProducts = useProducts();
+    const workTemplatesLinkedItems = useSelector((state: GlobalState) => getWorkTemplatesLinkedProducts(state));
     let tourSteps: Record<string, number> = TTNameMapToTourSteps[tourCategory];
 
+    let boardsProductEnabled = false;
+    let playbooksProductEnabled = false;
+    if (pluginProducts) {
+        boardsProductEnabled = pluginProducts.some((product) => (product.pluginId === suitePluginIds.focalboard) || (product.pluginId === suitePluginIds.boards));
+        playbooksProductEnabled = pluginProducts.some((product) => product.pluginId === suitePluginIds.playbooks);
+    }
```
Comment (marianunez) [inline components/tours/hooks.ts:52]: Wondering if we should move this check to a shareable selector as we may start repeating this in several places

File: components/tours/hooks.ts [hunk 2]
```diff
@@ -5,44 +5,72 @@ import {useCallback} from 'react';
...
+        if (!boardsPlugin && !boardsProductEnabled) {
+            delete steps.BOARDS_TOUR;
+        }
+        tourSteps = steps;
+    } else if (tourCategory === TutorialTourName.WORK_TEMPLATE_TUTORIAL) {
+        const steps: Record<string, number> = tourSteps as typeof WorkTemplateTourSteps;
+
+        if (workTemplatesLinkedItems.playbooks && workTemplatesLinkedItems.playbooks === 0) {
+            delete steps.PLAYBOOKS_TOUR;
+        }
```
Comment (marianunez) [inline components/tours/hooks.ts:71]: Could `workTemplatesLinkedItems.playbooks` ever be `null` and should we also delete the step in that case?

File: plugins/rhs_plugin/index.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    const pluggableId = getPluggableId(state);
+    const pluginComponent = rhsPlugins.find((element: PluginComponent) => element.id === pluggableId);
+    const pluginTitle = pluginComponent ? pluginComponent.title : '';
+
+    // work templates related code to show the tourtip
+    const activeRhsComponent = getActiveRhsComponent(state);
+    const pluginId = activeRhsComponent?.pluginId || '';
+
+    const currentUserId = getCurrentUserId(state);
+    const enableTutorial = state.entities.general.config.EnableTutorial === 'true';
```
Comment (marianunez) [inline plugins/rhs_plugin/index.ts:34]: I believe this setting is for tutorials related to the onboarding but in the work templates the tooltips are more of a response of the action of creating the work templates. Not sure if we should not show if based on this config. Thoughts @esethna?

File: plugins/rhs_plugin/index.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+    const tutorialStep = getInt(state, TutorialTourName.WORK_TEMPLATE_TUTORIAL, currentUserId, 0);
+
+    const workTemplateTourTipShown = tutorialStep === WorkTemplateTourSteps.FINISHED;
+    const showProductTour = !workTemplateTourTipShown && enableTutorial;
```
Comment (marianunez) [inline plugins/rhs_plugin/index.ts:39]: Could we move this logic up more towards the work template tooltip component? Would be nice to have this RHS_plugin as generic as possible

File: plugins/rhs_plugin/index.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+    const channelLinkedItems = showProductTour ? getWorkTemplatesLinkedProducts(state) : {};
+
+    const boardsCount = channelLinkedItems?.boards || 0;
+    const playbooksCount = channelLinkedItems?.playbooks || 0;
+
+    const showBoardsTour = showProductTour && pluginId === suitePluginIds.boards && boardsCount > 0;
+    const showPlaybooksTour = showProductTour && pluginId === suitePluginIds.playbooks && playbooksCount > 0;
```
Comment (marianunez) [inline plugins/rhs_plugin/index.ts:47]: Maybe same here as well, could we move all this checks into the `BoardsTourTip`/`PlaybooksTourTip` component so we can minimize the amount of additional checks here and have it as contained as possible in the specific work templates components? This way we just add the components without any checks and it can internally determine if needed to show

File: components/tours/onboarding_explore_tools_tour/explore_tools_tour_manager.tsx [hunk 6]
```diff
@@ -1,225 +0,0 @@
-// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Comment (marianunez) [inline components/tours/onboarding_explore_tools_tour/explore_tools_tour_manager.tsx:1]: Is this removal intended?

**Review by marianunez — CHANGES_REQUESTED** [review]
Thanks @pvev! Some comments below regarding trying to keep the `rhs_plugin` component as generic as possible. 
Also, great cleanup moving to TS in a few there 👍

File: components/tours/onboarding_explore_tools_tour/explore_tools_tour_manager.tsx (hunk 6 above)
Comment (pvev) [reply components/tours/onboarding_explore_tools_tour/explore_tools_tour_manager.tsx:1]: yes, no longer needed. It was a leftover after the refactoring made to the tourtips.

File: plugins/rhs_plugin/index.ts (hunk 5 above)
Comment (pvev) [reply plugins/rhs_plugin/index.ts:46]: done!

File: plugins/rhs_plugin/index.ts (hunk 4 above)
Comment (pvev) [reply plugins/rhs_plugin/index.ts:38]: done!

File: plugins/rhs_plugin/index.ts (hunk 3 above)
Comment (pvev) [reply plugins/rhs_plugin/index.ts:33]: not sure if could somehow eventually affect e2e, that's why I rather to keep it under that config value.

File: components/tours/hooks.ts (hunk 1 above)
Comment (pvev) [reply components/tours/hooks.ts:52]: done

File: components/global_header/left_controls/product_menu/product_menu.tsx [hunk 7]
```diff
@@ -80,10 +80,16 @@ const ProductMenu = (): JSX.Element => {
     const exploreToolsTourTriggered = triggerStep === GenericTaskSteps.STARTED;
 
     const pluginsList = useSelector((state: GlobalState) => state.plugins.plugins);
-    const boards = pluginsList.focalboard;
     const playbooks = pluginsList.playbooks;
 
-    const showBoardsTour = enableTutorial && tutorialStep === ExploreOtherToolsTourSteps.BOARDS_TOUR && exploreToolsTourTriggered && boards;
+    const focalboardPluginEnabled = pluginsList.focalboard;
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/global_header/left_controls/product_menu/product_menu.tsx:85]: done

File: components/global_header/left_controls/product_menu/product_menu.tsx [hunk 8]
```diff
@@ -80,10 +80,16 @@ const ProductMenu = (): JSX.Element => {
...
-    const boards = pluginsList.focalboard;
     const playbooks = pluginsList.playbooks;
 
-    const showBoardsTour = enableTutorial && tutorialStep === ExploreOtherToolsTourSteps.BOARDS_TOUR && exploreToolsTourTriggered && boards;
+    const boardsPluginEnabled = pluginsList.focalboard;
+    let boardsProductEnabled = false;
+    if (products) {
+        boardsProductEnabled = products.some((product) => product.pluginId === suitePluginIds.focalboard || product.pluginId === suitePluginIds.boards);
+    }
+    const boardsEnabled = boardsPluginEnabled || boardsProductEnabled;
```
Comment (marianunez) [inline components/global_header/left_controls/product_menu/product_menu.tsx:90]: Could we reuse your new `useGetPluginsActivationState` here?

File: components/tours/hooks.ts (hunk 2 above)
Comment (marianunez) [reply components/tours/hooks.ts:63]: Just to make sure I am following, if `workTemplatesLinkedItems.playbooks` is `null` that playbooks step was never added to begin with?

**Review by marianunez — APPROVED** [review]
LGTM, thanks @pvev! Non-blocking comments below

**Comment by marianunez** [comment]
Adding @neallred as another reviewer while Julien is out.

**Review by marianunez — COMMENTED** [review]
Also, are there any tests we can add to the `WorkTemplate_explore_tour`?

File: components/work_templates/index.tsx [hunk 9]
```diff
@@ -170,6 +191,85 @@ const WorkTemplateModal = () => {
...
+
+            // this one is for defining the auto tour start for the tour tip
+            {
+                user_id: currentUserId,
+                category: tourCategory,
+                name: TTNameMapToATStatusKey[tourCategory],
+                value: String(AutoTourStatus.ENABLED),
+            },
+        ];
+        await dispatch(savePreferences(currentUserId, preferences));
```
Comment (neallred) [inline components/work_templates/index.tsx:266]: Should the preferences definition and savePreferences call also go inside the `if (pluginId) {` check? Seems like if there were no boards or playbooks on the work template, we wouldn't want to re-show the tour.

File: components/work_templates/index.tsx [hunk 10]
```diff
@@ -170,6 +191,85 @@ const WorkTemplateModal = () => {
...
+
+        // store in the global state the plugins/integrations information related to the used template
+        // so we can display that data in the tourtip
+        await dispatch(onExecuteSuccess(linkedProductsCount));
```
Comment (neallred) [inline components/work_templates/index.tsx:270]: Is it important that `onExecuteSuccess` waits for `savePreferences` to complete? If not, I think we can get rid of the `async` on this function because nothing else within it waits, and it looks like where it is called, it is not awaited either.

File: i18n/en.json [hunk 11]
```diff
@@ -4383,6 +4383,14 @@
...
   "plan.cloud": "Cloud",
   "plan.self_serve": "Self-serve",
+  "pluggable_rhs.tourtip.boards.access": "Access your linked boards from the Boards icon on the right hand App bar.",
+  "pluggable_rhs.tourtip.boards.click": "Click into boards from this right panel.",
+  "pluggable_rhs.tourtip.boards.review": "Review board updates from your channels.",
+  "pluggable_rhs.tourtip.boards.title": "Access your {count} linked boards!",
+  "pluggable_rhs.tourtip.playbooks.access": "Access your linked playbooks from the Playbooks icon on the right hand App bar.",
+  "pluggable_rhs.tourtip.playbooks.click": "Click into playbooks from this right panel.",
+  "pluggable_rhs.tourtip.playbooks.review": "Review playbook updates from your channels.",
+  "pluggable_rhs.tourtip.playbooks.title": "Access your {count} linked playbook run.",
```
Comment (neallred) [inline i18n/en.json:4393]: Nit: pluralization between this and `"pluggable_rhs.tourtip.boards.title": "Access your {count} linked boards!",` is inconsistent. I don't care how we choose to do the pluralization, but we should probably be consistent since these will be seen close together.

File: packages/components/src/tour_tip/tour_tip_backdrop.tsx [hunk 12]
```diff
@@ -54,7 +56,7 @@ export const TourTipBackdrop = ({
     const backdrop = (
         <div
             onClick={onDismiss}
-            className={'tour-tip__backdrop'}
+            className={`tour-tip__backdrop${transparent ? '_transparent' : ''}`}
```
Comment (neallred) [inline packages/components/src/tour_tip/tour_tip_backdrop.tsx:59]: Since this classname is doing a BEM convention, I'd recommend following it for the modifier and going `--transparent`.

File: packages/mattermost-redux/src/reducers/entities/work_templates.ts [hunk 13]
```diff
@@ -57,9 +57,23 @@ function playbookTemplates(state: [] = [], action: GenericAction) {
     }
 }
 
+function linkedProducts(state: Record<string, number> = {}, action: GenericAction) {
+    switch (action.type) {
+    case WorkTemplatesType.EXECUTE_SUCCESS: {
+        return {
```
Comment (neallred) [inline packages/mattermost-redux/src/reducers/entities/work_templates.ts:63]: Since the state we're modelling is the counts of what was linked in the work template that was just executed, I think it would be a little more robust to `return action.data` instead of merging it with the existing state. If the implementation of `getContentCount` were altered in the future so that only items with a positive count were in the object, or other keys were only conditionally there, then you could get stale data where a first work template run that had boards and the second one did not have boards, you would keep seeing the boards UI even though it was relevant to the second work template execution.

File: packages/mattermost-redux/src/selectors/entities/general.ts [hunk 14]
```diff
@@ -113,3 +113,5 @@ export const isMarketplaceEnabled: (state: GlobalState) => boolean = createSelec
         return config.PluginsEnabled === 'true' && config.EnableMarketplace === 'true';
     },
 );
+
+export const getWorkTemplatesLinkedProducts = (state: GlobalState) => state.entities.worktemplates.linkedProducts;
```
Comment (neallred) [inline packages/mattermost-redux/src/selectors/entities/general.ts:117]: :+1: for adding a selector. I think it is worth it to create a `packages/mattermost-redux/src/selectors/entities/worktemplates.ts` file since the state is somewhat complex, will likely grow more complex over time given the potential of this feature, and there are some existing ad-hoc selectors in `components/work_templates/index.tsx` that we will eventually also want to place in a selectors file.

File: plugins/rhs_plugin/index.ts [hunk 15]
```diff
@@ -0,0 +1,26 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Comment (neallred) [inline plugins/rhs_plugin/index.ts:1]: :100:

File: plugins/rhs_plugin/rhs_plugin.tsx [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+import {BoardsTourTip, PlaybooksTourTip} from 'components/tours/worktemplate_explore_tour';
+
+import Pluggable from 'plugins/pluggable';
+
+import AutoShowLinkedBoardTourTip from './auto_show_linked_board_tourtip';
+
+export type Props = {
+    showPluggable: boolean;
+    pluggableId: string;
+    title: string | React.ReactNode;
```
Comment (neallred) [inline plugins/rhs_plugin/rhs_plugin.tsx:16]: Nit: Should be able to omit `string` since it is included in `React.ReactNode;`

**Review by neallred — APPROVED** [review]
Couldn't quite figure out how to get worktemplates working locally, so I haven't run it, but it LGTM from a code perspective! There's some nits and suggestions, but none of them are blocking.

File: plugins/rhs_plugin/index.ts (hunk 3 above)
Comment (esethna) [reply plugins/rhs_plugin/index.ts]: +1 to Pablo, let's hide any tips or tutorials behind this flag for e2e purposes

File: components/tours/hooks.ts (hunk 2 above)
Comment (pvev) [reply components/tours/hooks.ts:63]: on initial render workTemplatesLinkedItems is an empty object (https://github.com/mattermost/mattermost-webapp/pull/12219/files#diff-e75490b587f9fa072658f9d6b223817f8ad54af3f1c4716335b41d40e0b800fbR262), and it will keep that way until the executeTemplate is executed sucessfully, then the information about the items used by the template is summarized and filled in the workTemplatesLinkedItems object via a reducer placed in the global state (https://github.com/mattermost/mattermost-webapp/pull/12219/files#diff-14ff19a2b845055ffa501bef257b8f9ad99942f84dc252b6adef5b48f0e96b6aR270). So, to avoid initially deleting the step from the constant already loaded variable in memory, we first check that in that particular condition of the tourtip, for starting the information must be populated in the global state, specially before verifying the quantity and if the quantity is 0 then proceed to delete that step. Summarizing, in this particular scenario (the worktemplates tourtip), the initial global state is empty and we know and assume that on the subsequent render that value must have been populated before proceeding with the logic.

File: plugins/rhs_plugin/rhs_plugin.tsx (hunk 16 above)
Comment (pvev) [reply plugins/rhs_plugin/rhs_plugin.tsx:16]: thanks, done!

File: packages/mattermost-redux/src/selectors/entities/general.ts (hunk 14 above)
Comment (pvev) [reply packages/mattermost-redux/src/selectors/entities/general.ts:117]: agree that maybe it would be worth in the future, but for now let's keep it simple since we are not sure the future roadmap of this feature.

File: packages/mattermost-redux/src/reducers/entities/work_templates.ts (hunk 13 above)
Comment (pvev) [reply packages/mattermost-redux/src/reducers/entities/work_templates.ts:63]: the way it is implemented right now is that in every successful execution that data model is overwritten, so on every other worktemplate use, the data in the global store contains the information for the new template included items.

File: packages/components/src/tour_tip/tour_tip_backdrop.tsx (hunk 12 above)
Comment (pvev) [reply packages/components/src/tour_tip/tour_tip_backdrop.tsx:59]: thanks, done

File: i18n/en.json (hunk 11 above)
Comment (pvev) [reply i18n/en.json:4393]: nice catch, thanks! done!

File: components/work_templates/index.tsx (hunk 10 above)
Comment (pvev) [reply components/work_templates/index.tsx:270]: right, no need to await for it

File: components/work_templates/index.tsx [hunk 17]
```diff
@@ -170,6 +191,85 @@ const WorkTemplateModal = () => {
         setSelectedVisibility(visibility);
     };
 
+    const getContentCount = (template: WorkTemplate) => {
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/work_templates/index.tsx:194]: thanks, done

File: components/work_templates/hooks.ts [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+
+import {useSelector} from 'react-redux';
+
+import {GlobalState} from 'types/store';
+import {PluginComponent} from 'types/store/plugins';
+
+type HookReturnType = {
+    pluggableId: string;
+    rhsPluggableIds: Map<string, string>;
+    pluginComponent: PluginComponent | undefined;
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/work_templates/hooks.ts:12]: thanks, done

File: components/tours/worktemplate_explore_tour/useShowTourTip.tsx [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+import {TutorialTourName, WorkTemplateTourSteps} from '../constant';
+
+import {GlobalState} from 'types/store';
+
+export const useShowShowTourTip = () => {
+    const activeRhsComponent = useSelector((state: GlobalState) => getActiveRhsComponent(state));
+    const pluginId = activeRhsComponent?.pluginId || '';
+
+    const currentUserId = useSelector((state: GlobalState) => getCurrentUserId(state));
+    const enableTutorial = useSelector((state: GlobalState) => state.entities.general.config.EnableTutorial === 'true');
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/tours/worktemplate_explore_tour/useShowTourTip.tsx:21]: thanks, done

File: components/tours/worktemplate_explore_tour/playbooks_tour_tip.tsx [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+            <li>
+                {formatMessage({
+                    id: 'pluggable_rhs.tourtip.playbooks.review',
+                    defaultMessage: 'Review playbook updates from your channels.',
+                })}
+            </li>
+        </ul>
+    );
+
+    if (!showPlaybooksTour) {
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/tours/worktemplate_explore_tour/playbooks_tour_tip.tsx:45]: thanks, done!

File: components/tours/worktemplate_explore_tour/playbooks_tour_tip.tsx [hunk 21]
```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React from 'react';
+import {FormattedMessage, useIntl} from 'react-intl';
+
+import {useShowShowTourTip} from './useShowTourTip';
+import OnboardingWorkTemplateTourTip from './worktemplate_explore_tour_tip';
+
+export const PlaybooksTourTip = (): JSX.Element | null => {
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/tours/worktemplate_explore_tour/playbooks_tour_tip.tsx:10]: indeed ReactNode contains several types including null but I find it not as descriptive as mentioning in the return type that for a particular condition the component might be null. But my preference for JSX.Element goes more as mentioned before as a personal preference. Some authors also suggest to use ReactElement that at the end is the same as JSX.Element but implemented for React.

File: components/tours/worktemplate_explore_tour/useShowTourTip.tsx [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+import {getWorkTemplatesLinkedProducts} from 'mattermost-redux/selectors/entities/general';
+import {getInt} from 'mattermost-redux/selectors/entities/preferences';
+
+import {getActiveRhsComponent} from 'selectors/rhs';
+import {suitePluginIds} from 'utils/constants';
+import {TutorialTourName, WorkTemplateTourSteps} from '../constant';
+
+import {GlobalState} from 'types/store';
+
+export const useShowShowTourTip = () => {
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/tours/worktemplate_explore_tour/useShowTourTip.tsx:16]: thanks, done

File: components/tours/tour_manager.tsx [hunk 23]
```diff
@@ -69,7 +71,7 @@ export const useTourTipManager = (tourCategory: string): ChannelsTourTipManager
                 trackEventAction(tourCategory, telemetryTag);
             }
         },
-        [currentUserId],
+        [currentUserId, currentChannelId],
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/tours/tour_manager.tsx:74]: thanks, done

File: components/tours/hooks.ts [hunk 24]
```diff
@@ -5,44 +5,64 @@ import {useCallback} from 'react';
...
     TTNameMapToTourSteps,
     TutorialTourName,
+    WorkTemplateTourSteps,
 } from './constant';
 
 export const useGetTourSteps = (tourCategory: string) => {
     const isGuestUser = useSelector((state: GlobalState) => isCurrentUserGuestUser(state));
-    const pluginsList = useSelector((state: GlobalState) => state.plugins.plugins);
 
+    const workTemplatesLinkedItems = useSelector((state: GlobalState) => getWorkTemplatesLinkedProducts(state));
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/tours/hooks.ts:43]: thanks, done

File: components/global_header/left_controls/product_menu/product_menu.tsx (hunk 8 above)
Comment (pvev) [reply components/global_header/left_controls/product_menu/product_menu.tsx]: thanks, done

**Review by stevemudie — APPROVED** [review]
Can't get the Spinwick to go so I'll test this post merge.

File: components/work_templates/utils.ts [hunk 25]
```diff
@@ -31,3 +32,32 @@ export function getTemplateDefaultIllustration(template: WorkTemplate): string {
     return '';
 }
 
+export const getContentCount = (template: WorkTemplate, playbookTemplates: PlaybookTemplateType[]) => {
+    const res = {
+        playbooks: 0,
+        boards: 0,
+        jira: 0,
```
Comment (JulienTant) [inline components/work_templates/utils.ts:39]: nit: given that we don't really point at those ones, this seems unnecessary.

**Review by JulienTant — APPROVED** [review]
LGTM @pvev

**Comment by amyblais** [comment]
@pvev Please help open a manual cherry-pick PR for the ``release-7.9`` branch.

**Comment by pvev** [comment]
@amyblais I can open a manual cherry-pick for this one and this one should also be cherry-picked https://github.com/mattermost/mattermost-webapp/pull/12301
cc: @marianunez

**Comment by spirosoik** [comment]
I am doing some testing
